core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1764,11 +1764,14 @@ private[zio] trait ZIOFunctions extends Serializable {
     n: Long
   )(as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
     (for {
-      semaphore <- Semaphore.make(n)
-      bs <- foreachPar[R, E, A, B](as) { a =>
-             semaphore.withPermit(fn(a))
-           }
-    } yield bs).refailWithTrace
+      q     <- Queue.bounded[(Promise[E, B], A)](n.toInt)
+      pairs <- ZIO.foreach(as)(a => Promise.make[E, B].map(p => (p, a)))
+      _     <- ZIO.foreach(pairs)(pair => q.offer(pair)).fork
+      _ <- ZIO.collectAll(List.fill(n.toInt)(q.take.flatMap {
```

Comment:
Instead of filling a `List`, we could do this in one go. It's not easy now, but I will add a ticket for `ZIO.fill`.

Reply:
Sounds good. I would be happy to pick up that ticket next.